File: attractors/2D/Pygame/Bedhead.py
#
#  (C) Copyright 2024  Pavel Tisnovsky
#
#  All rights reserved. This program and the accompanying materials
#  are made available under the terms of the Eclipse Public License v1.0
#  which accompanies this distribution, and is available at
#  http://www.eclipse.org/legal/epl-v10.html
#
#  Contributors:
#      Pavel Tisnovsky
#


from math import cos, sin
from typing import Tuple

from renderer import render_attractor
from ui import event_loop, initialize

# window settings
WINDOW_WIDTH = 800
WINDOW_HEIGHT = 600
WINDOW_TITLE = "Bedhead dynamical system"


def bedhead(x: float, y: float, A: float, B: float) -> Tuple[float, float]:
    """Calculate next point in the strange attractor."""
    x_dot = sin(x * y / B) * y + cos(A * x - y)
    y_dot = x + sin(y) / B
    return x_dot, y_dot


def main() -> None:
    # initialize user interface based on Pygame
    display, surface, clock = initialize(WINDOW_TITLE, WINDOW_WIDTH, WINDOW_HEIGHT)

    # redraw the whole dynamical system
    render_attractor(
        surface=surface,
        x=0,
        y=0,
        scale=80.0,
        x_offset=0,
        y_offset=40,
        max_points=1000000,
        settle_down_points=10,
        contrast=1 / 1000.0,
        attractor_formula=bedhead,
        filename="bedhead.png",
        A=0.06,
        B=0.98,
    )

    # and enter the event loop
    event_loop(display, surface, clock)


if __name__ == "__main__":
    main()

File: attractors/2D/Go/Bedhead.go
/*
 *  (C) Copyright 2025  Pavel Tisnovsky
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  which accompanies this distribution, and is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 *
 *  Contributors:
 *      Pavel Tisnovsky
 */

package main

import (
	"attractors/bitmap"
	"image"
	"math"
)

const ImageWidth = 800
const ImageHeight = 600
const Maxiter = 1000000

func bedhead(x, y, a, b float64) (float64, float64) {
	xn := math.Sin(x*y/b)*y + math.Cos(a*x-y)
	yn := x + math.Sin(y)/b
	return xn, yn
}

func calcSystem(
	width, height, maxiter int,
	A, B, scale, xOffset, yOffset, maxFactor float64) *image.RGBA {

	img := image.NewRGBA(image.Rect(0, 0, width, height))

	bmp := bitmap.CreateBitmap(width, height)
	maxPoints := maxiter
	settleDownPoints := 100

	x := 0.0
	y := 0.0

	for i := range maxPoints {
		xn, yn := bedhead(x, y, A, B)
		xi := int(float64(width)/2 + scale*xn + xOffset)
		yi := int(float64(height)/2 + scale*yn + yOffset)
		if i > settleDownPoints {
			if xi >= 0 && yi >= 0 && xi < int(width) && yi < int(height) {
				bmp[yi][xi] += 1.0
			}
		}
		x, y = xn, yn
	}

	bitmap.DrawBitmapToImage(bmp, img, maxFactor)
	return img
}

func main() {
	img := calcSystem(ImageWidth, ImageHeight, Maxiter, 0.06, 0.98, 80, 50, 10, 1/100.0)
	bitmap.SaveImage(img, "bedhead_1.png")

	img = calcSystem(ImageWidth, ImageHeight, Maxiter, -0.67, 0.83, 220, -100, -350, 1/100.0)
	bitmap.SaveImage(img, "bedhead_2.png")

	img = calcSystem(ImageWidth, ImageHeight, Maxiter, -0.67, 0.68, 70, 0, 50, 1/50.0)
	bitmap.SaveImage(img, "bedhead_3.png")
}
